### PR TITLE
Data Module: Support controls in resolvers

### PR DESCRIPTION
**Pull request by youknowriad**
This PR tweaks the data module to allow extended usage of the controls API in resolvers. To achieve that it does:

 - Refactors the way resolvers are registered and fulfilled in the data module
 - It exposes a `fulfill` function in the registry essentially to allow plugins to override the fulfillment implementation.
 - Refactors the redux-routine middleware to allow the possibility to know when the execution of the routine ends
 - Extend the `fulfill` function in the controls plugin to support resolvers in the controls plugin.

**Testing instructions**

 - Check that the current resolvers work properly (we don't use the controls plugin yet in Gutenberg)

### CODE REVIEW

File: packages/redux-routine/README.md [hunk 1]
```diff
@@ -22,9 +22,9 @@ For example, consider a common case where we need to issue a network request. We
 
 ```js
 import { combineReducers, createStore, applyMiddleware } from 'redux';
-import createRoutineMiddleware from '@wordpress/redux-routine';
+import createMiddleware from '@wordpress/redux-routine';
```
Comment (aduth) [inline packages/redux-routine/README.md]: There is no default export anymore?

File: packages/redux-routine/README.md [hunk 2]
```diff
@@ -57,7 +57,7 @@ request has completed does the action creator procede to return the `SET_TEMPERA
 
 ## API
 
-### `createMiddleware( controls: ?Object )`
+### `createMiddleware( runtime: function )`
```
Comment (aduth) [inline packages/redux-routine/README.md]: Did this actually change? The function signature looks the same.

Comment (youknowriad) [reply packages/redux-routine/README.md]: I changed and reverted :P

File: packages/data/src/plugins/controls/index.js [hunk 3]
```diff
@@ -26,5 +27,14 @@ export default function( registry ) {
 
 			return store;
 		},
+
+		fulfill( reducerKey ) {
+			return mapValues(
+				get( registry.namespaces, [ reducerKey, 'resolvers' ] ),
```
Comment (aduth) [inline packages/data/src/plugins/controls/index.js]: When will `registry.namespaces` be falsey

File: packages/data/src/plugins/controls/index.js [hunk 4]
```diff
@@ -26,5 +27,14 @@ export default function( registry ) {
 
 			return store;
 		},
+
+		fulfill( reducerKey ) {
```
Comment (aduth) [inline packages/data/src/plugins/controls/index.js]: To me, it seems more sensible and easy-to-extend to accept as arguments the full thing of the thing we want to fulfill: with selector name and arguments.

Comment (youknowriad) [reply packages/data/src/plugins/controls/index.js]: The idea was to mimic the "select" and "dispatch" APIs but I hesitated a lot. I don't mind changing, what do you think?

Comment (youknowriad) [reply packages/data/src/plugins/controls/index.js]: On a second thought, this doesn't really make sense as a public API aside for plugins, which means we don't need to mimic the select/dispatch behavior. I'm going to update.

File: packages/redux-routine/README.md (hunk 1 above)
Comment (youknowriad) [reply packages/redux-routine/README.md]: Fixed

**Review by aduth — COMMENTED** [review]
`fulfill` is fine, `rungen` is fine, promises returned from dispatch is fine.

I'm not really digging our mixed consideration of resolvers. Is a resolver one of the proper concepts of the registry? If so, we're losing optimizations by wrapping selectors in `registerResolvers`. If not, we need to stop referencing `resolvers` in the plugins, in `registerSelectors`, and in the registry `fulfill`.

File: packages/data/src/middlewares/async-generator.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * Normalizes the given object argument to an async iterable, asynchronously
+ * yielding on a singular or array of generator yields or promise resolution.
+ *
+ * @param {*} object Object to normalize.
+ *
+ * @return {AsyncGenerator} Async iterable actions.
+ */
+export function toAsyncIterable( object ) {
```
Comment (aduth) [inline packages/data/src/middlewares/async-generator.js]: Not sure if we want to do it here, but since we're "starting fresh" with this, might be an opportunity to implement some simplifications that @jsnajdr had suggested in #8913.

File: packages/data/src/middlewares/async-generator.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+/**
+ * Simplest possible promise redux middleware.
+ *
+ * @param {Object} store Redux store.
+ *
+ * @return {function} middleware.
+ */
+const asyncGeneratorMiddleware = ( store ) => ( next ) => ( action ) => {
+	if ( ! isAsyncIterable( action ) ) {
+		next( action );
```
Comment (aduth) [inline packages/data/src/middlewares/async-generator.js]: Is this meant to `return` ?

File: packages/data/src/plugins/async-generator/index.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+		},
+
+		async fulfill( reducerKey, selectorName, ...args ) {
+			const resolver = get( registry.namespaces, [ reducerKey, 'resolvers', selectorName ] );
+			if ( ! resolver ) {
+				return;
+			}
+			const store = registry.namespaces[ reducerKey ].store;
+			const state = store.getState();
+			const actionCreator = resolver.fulfill( state, ...args );
```
Comment (aduth) [inline packages/data/src/plugins/async-generator/index.js]: Would `fulfill` (the function, not the result) be considered the action creator? And what we're assigning to be the `action` ?

File: packages/data/src/plugins/async-generator/index.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+			if ( actionCreator ) {
+				await store.dispatch( toAsyncIterable( actionCreator ) );
```
Comment (aduth) [inline packages/data/src/plugins/async-generator/index.js]: So the idea with the stock middlewares (promise, async generator) is that they're expected to be present for use by plugins? Would it not be possible for the plugins themselves to ensure their presence if needed? Via middleware application in the overrided `registerStore` ?

File: packages/data/src/registry.js [hunk 9]
```diff
@@ -347,6 +273,7 @@ export function createRegistry( storeConfigs = {} ) {
 	}
 
 	let registry = {
+		namespaces,
```
Comment (aduth) [inline packages/data/src/registry.js]: I recall in early iterations of plugins implementation I needed to consider in `withPlugins` whether the result of `mapValues` should be left alone if not a function. Here, `namespaces` is being added and is unique from the others in that it is an object, not a function. It cannot therefore be `apply`'d.

File: test/unit/__mocks__/@wordpress/data.js [hunk 10]
```diff
@@ -0,0 +1,3 @@
+import { use, plugins } from '../../../../packages/data/src';
+use( plugins.asyncGenerator );
```
Comment (aduth) [inline test/unit/__mocks__/@wordpress/data.js]: What thing is depending on async generators in the test? Should they be, or should those tests be part of the suite for the `asyncGenerator` plugin instead?

File: packages/data/src/middlewares/async-generator.js (hunk 5 above)
Comment (youknowriad) [reply packages/data/src/middlewares/async-generator.js]: Well I'm hoping we remove it actually in two releases or so, so might not be worth it.

File: packages/data/src/plugins/async-generator/index.js (hunk 7 above)
Comment (youknowriad) [reply packages/data/src/plugins/async-generator/index.js]: The action creator is `resolver.fulfill` yes and I'm not sure I understand the second part of the question.

In this particular case though (async generator), we can't consider a resolver as an action creator because the resolvers takes the state as first argument (and action creators don't) and that's exactly why I'm overriding the default `fulfill` function of the registry.

File: packages/data/src/plugins/async-generator/index.js (hunk 8 above)
Comment (youknowriad) [reply packages/data/src/plugins/async-generator/index.js]: Isn't it the case already? 

Aside the `promise` middleware which is considered the "core" implementation of the resolvers and that's why it's not in a plugin. The others are already.

File: test/unit/__mocks__/@wordpress/data.js (hunk 10 above)
Comment (youknowriad) [reply test/unit/__mocks__/@wordpress/data.js]: This is needed for some `effects` tests. We have some effects triggering async generator resolvers that won't work without this.

Comment (youknowriad) [reply test/unit/__mocks__/@wordpress/data.js]: This is something that won't be needed in the future if we rewrite those effects as generators with controls as we can directly feed the generator with mocked results in the tests without running the controls.

File: packages/data/src/middlewares/async-generator.js (hunk 6 above)
Comment (youknowriad) [reply packages/data/src/middlewares/async-generator.js]: This fixed the warnings I was seeing. Thanks :)

**Comment by youknowriad** [comment]
Ok, this PR now is in a much better shape. I think it's getting ready to 🚢 

I wanted to deprecated the async-generator middleware/plugin but thought I would first wait for us to refactor the current async generators we have in Gutenberg to use controls and do that right after.

File: packages/data/src/registry.js [hunk 11]
```diff
@@ -167,8 +166,37 @@ export function createRegistry( storeConfigs = {} ) {
 	 */
 	function registerSelectors( reducerKey, newSelectors ) {
 		const store = namespaces[ reducerKey ].store;
-		const createStateSelector = ( selector ) => ( ...args ) => selector( store.getState(), ...args );
-		namespaces[ reducerKey ].selectors = mapValues( newSelectors, createStateSelector );
+		const createStateSelector = ( selector, selectorName ) => ( ...args ) => {
```
[the remark this answers is not in the document]
Comment (aduth) [reply packages/data/src/registry.js]: I think this is still important to address.

File: packages/data/src/registry.js [hunk 12]
```diff
@@ -336,17 +262,21 @@ export function createRegistry( storeConfigs = {} ) {
...
-	 * @param {Object<string,Function>} fns Object of function values.
+	 * @param {Object<string,Function>} attributes Object of function values.
 	 *
 	 * @return {Object<string,Function>} Object enhanced with plugin proxying.
 	 */
-	function withPlugins( fns ) {
-		return mapValues( fns, ( fn, key ) => function() {
+	function withPlugins( attributes ) {
+		return mapValues( attributes, ( attribute, key ) => function() {
+			if ( key === 'attributes' ) {
```
Comment (aduth) [inline packages/data/src/registry.js]: What is `attributes` ? Is this meant to be `namespaces` ? Or better, could we just have this be an `typeof key !== 'function'` condition?

File: packages/data/src/plugins/controls/index.js [hunk 13]
```diff
@@ -22,9 +23,24 @@ export default function( registry ) {
...
 			return store;
 		},
+
+		async fulfill( reducerKey, selectorName, ...args ) {
+			if ( ! registry.namespaces[ reducerKey ].supportControls ) {
+				registry.fulfill( reducerKey, selectorName, ...args );
+				return;
+			}
+
+			const resolver = get( registry.namespaces, [ reducerKey, 'resolvers', selectorName ] );
```
[the remark this answers is not in the document]
Comment (aduth) [reply packages/data/src/plugins/controls/index.js]: I was under the impression that part of the intent with introducing the promise middleware was to make it so that at the very least, `controls` doesn't need to know of resolvers, and at best `controls` doesn't even need to implement `fulfill` at all.

Why again did we add the promise middleware? Just to refactor async generators?

[the remark this answers is not in the document]
Comment (youknowriad) [reply packages/data/src/plugins/controls/index.js]: I'm pretty sure I explained this elsewhere :)

The promise middleware is a way to support resolvers without any plugins:

 - A resolver is action creator returning a promise

And yes, we don't need `fulfill` to implement `controls` support for resolvers BUT we do need it temporarily to support "async-generators" because async generators (like we write them today) can't be considered action creators because they take "state" as first argument, which means I needed a way to override the default `fulfill` for async generators. And by doing so, I needed to restore the default `fulfill` behavior for controls.

In other words:

 - It's temporary
 - When we remove the async-generators plugin, we will remove entirely the `fulfill` and `namespaces` from the registry object and we will also remove the `fulfill` defined in the controls plugin.

[the remark this answers is not in the document]
Comment (aduth) [reply packages/data/src/plugins/controls/index.js]: Should we be deprecating `async-generator` then? Now? In a follow-up pull request?

[the remark this answers is not in the document]
Comment (youknowriad) [reply packages/data/src/plugins/controls/index.js]: yes 

> I wanted to deprecated the async-generator middleware/plugin but thought I would first wait for us to refactor the current async generators we have in Gutenberg to use controls and do that right after.

File: packages/data/src/plugins/async-generator/index.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
+/**
+ * External dependencies
+ */
+import { applyMiddleware } from 'redux';
+import { get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import asyncGeneratorMiddleware, { toAsyncIterable } from '../../middlewares/async-generator';
```
Comment (aduth) [inline packages/data/src/plugins/async-generator/index.js]: Minor: To me, it seems more important to keep plugins self-contained than to keep middlewares together, i.e. middleware defined within `src/plugins/async-generator`

File: packages/data/src/registry.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (aduth) [reply packages/data/src/registry.js]: >I think this is still important to address.

I still think this is still important to address.

**Review by aduth — DISMISSED** [review]

File: packages/data/src/plugins/controls/index.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (aduth) [reply packages/data/src/plugins/controls/index.js]: >we will remove entirely the fulfill and namespaces from the registry object and we will also remove the fulfill defined in the controls plugin.

Should we mark as experimental then, to avoid need for deprecation?

File: packages/data/src/plugins/async-generator/index.js (hunk 14 above)
Comment (youknowriad) [reply packages/data/src/plugins/async-generator/index.js]: Where would you put the "promise" middleware?

File: packages/redux-routine/src/index.js [hunk 15]
```diff
@@ -12,42 +12,19 @@ import castError from './cast-error';
  * value of the yield assignment. If the control handler returns undefined, the
  * execution is not continued.
  *
- * @param {Object} controls Object of control handlers.
+ * @param {Object}    controls Object of control handlers.
```
Comment (aduth) [inline packages/redux-routine/src/index.js]: Minor: We don't want this change. There's no intent to align `@param` descriptions with `@return`.

File: packages/data/src/plugins/async-generator/index.js (hunk 14 above)
Comment (aduth) [reply packages/data/src/plugins/async-generator/index.js]: I don't really feel strongly where, either inline at `async-generator/index.js` or a separate `async-generator/middleware.js` file.

Comment (youknowriad) [reply packages/data/src/plugins/async-generator/index.js]: The "promise" middleware is not a plugin middleware it's "core". It will be kept even if the async-generator plugin is removed.

File: packages/redux-routine/src/runtime.js [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Internal dependencies
+ */
+import castError from './cast-error';
+
+/**
+ * Create a co-routine runtime.
+ *
+ * @param {Object}    controls Object of control handlers.
+ * @param {function}  dispatch Unhandled action dispatch.
```
[the remark this answers is not in the document]
Comment (aduth) [reply packages/redux-routine/src/runtime.js]: We should settle what this should be at some point. My previous comment is nonsense because `typeof {} === 'object'` (lowercase), yet we still use `Object` (uppercase) to describe the type.

Either to be discussed in Core JS or at least search through prior discussion to find what basis we had settled on here.

I'm thinking "Function" is correct though.

**Review by aduth — APPROVED** [review]
Consider `__experimental` comment I left at https://github.com/WordPress/gutenberg/pull/9507#discussion_r216729492

Otherwise LGTM 👍 Thanks for putting up with me 😄
